Make VNC and httpd services optional for docker image

jchristi opened the pull request:
My team does not use the VNC or httpd capabilities of the default docker image, which allows us to use the official [selenium docker images](https://hub.docker.com/r/selenium/standalone-firefox/tags/) that do not include VNC or httpd which make the images lighter weight.

This PR makes VNC and httpd optional by adding two environment variables: `WEBDRIVER_WHARF_IMAGE_VNC_ENABLED` and `WEBDRIVER_WHARF_IMAGE_HTTPD_ENABLED` which both default to `True` resulting in no change in default behavior. If either are set to `False`, then the associated docker ports will not be forwarded.

Lastly, this PR adds a `.travis.yml` file and a couple `make` targets for basic python syntax checking. This can be extended to add unit tests in the future.

seandst commented:
Please limit PRs to one topic. I have comments about the travis and make targets, but I'd prefer to address those in an unrelated PR and have this just be focused on the optional VNC/HTTP feature.

seandst commented on webdriver_wharf/app.py:
```diff
@@ -15,13 +15,24 @@
+# Set to True if the container runs a VNC service
+vnc_enabled = os.environ.get('WEBDRIVER_WHARF_IMAGE_VNC_ENABLED', True)
+# Set to True if the container runs an httpd service
+httpd_enabled = os.environ.get('WEBDRIVER_WHARF_IMAGE_HTTPD_ENABLED', True)
```
For all occurances: http, not httpd. HTTP is the generic service name, httpd is apache's web server.

jchristi replied:
I also forgot to convert from string to boolean :|

seandst commented on webdriver_wharf/app.py:
```diff
@@ -172,18 +183,21 @@ def container_info(container):
     }
+    if vnc_enabled:
+        container_info['vnc_port'] = container.vnc_port
+        container_info['vnc_display'] = 'vnc://%s:%d' % (host_noport, container.vnc_port - 5900)
```
Removing these from the container info is a backward-incompatible API change. While the API isn't well-defined, the docs (index page) at least guarantee that you will get a value for vnc_port and vnc_display, so they should be present in the container info, even if they're None.

I don't think this applies to `http_port`/`fileviewer_url`, since those are undocumented and at the moment I don't even think they work.

jchristi replied:
I've modified index_document variable to conditionally display VNC information.

seandst replied:
The problem isn't that vnc_display was in docs, the problem is that *because* vnc_display was in the docs, it's part of wharf's interface and must remain.

jchristi replied:
Suggestions? Can I return the keys but make the fields an empty string?

seandst replied:
I think what I said in the first comment is still good:

>  they should be present in the container info, even if they're None

The docs say that you will get a port and url from those values, but I think I'm painted into a corner on that one, since I love and want this change. :)

Probably rewording the docs to indicate that the values might be None (`null` in JSON) would be nice, and I'll double check that the CFME QE automation will be just fine with the change.

seandst commented on webdriver_wharf/db.py:
```diff
@@ -10,6 +10,8 @@
 from sqlalchemy.orm import sessionmaker
 from sqlalchemy.ext.declarative import declarative_base
 
+from app import vnc_enabled, httpd_enabled
```
This presents a small problem. app import db, and now db imports app. I'd  like to keep the import graph straightforward, so if possible I'd rather not see db import from app, and instead alter `interactions` to allow the corresponding values in `port_bindings` to be None, signaling that they shouldn't be forwarded.

The current port_bindings behavior was too tightly coupled with `interactions` even before this change, and it's possible that the correct solution is for this property to go away entirely, and have the port-forwarding decisions all be made in `interactions`.

jchristi replied:
How about a new file specifically for environment variables? Doesn't solve the larger problems but for purposes of this PR it would do the trick of resolving the import graph issue.

seandst commented:
With regard to the travis and mark target stuff, I've built on that in a personal feature branch, and I can open that up in another PR so you can check it out if you'd like. Let me know.

seandst commented:
This is a pain in the butt, but any new environment vars should be added to the readme. The likelihood that I've forgotten to do this myself is high.

jchristi commented:
Changed:
 * Removed the `.travis.yml` and the changes to the `Makefile`.
 * Added ENV variables to `README.md`
 * Moved ENV variables to `env.py` to simplify import graph
 * modified index page to only display VNC information if `WEBDRIVER_WHARF_IMAGE_VNC_ENABLED` is `True`
 * Changed occurences of "httpd" to "http"

seandst commented on webdriver_wharf/env.py:
```diff
@@ -0,0 +1,29 @@
+
+# Set to True if the container runs an http service
+http_enabled = str2bool(os.environ.get('WEBDRIVER_WHARF_IMAGE_HTTP_ENABLED',
+    True))
```
These don't need to wrap. The project style is set to 100 column line length in setup.cfg (use flake8, not pep8)

jchristi replied:
good to know. I guess my vim is automatically configured to pep8.

RonnyPfannschmidt commented:
the new plan is to exclusively use container expose and labels on containers to move the responsibility of ports and port properties to the container
